scipy/interpolate/_polyint.py

```diff
@@ -677,8 +742,91 @@ def _evaluate(self, x):
+            Derivative interpolated at the x-points. Shape of d is
+            determined by replacing the interpolation axis in the
+            original array with the shape of x.
+        """
+        x, x_shape = self._prepare_x(x)
+        y = self._evaluate_derivatives(x, der+1, all_lower=False)
+        return self._finish_y(y, x_shape)
+
+    def _evaluate_derivatives(self, x, der=None, all_lower=True):
+        # NB: der here is not the order of the highest derivative;
+        # instead, it is the size of the derivatives matrix that
+        # would be returned with all_lower=True, including the
+        # '0th' derivative (the undifferentiated function).
+        # E.g. to evaluate the 5th derivative alone, call
+        # _evaluate_derivatives(x, der=6, all_lower=False).
+
+        if (not all_lower) and (x.size == 0 or self.r == 0):
+            return np.zeros((0, self.r), dtype=self.dtype)
+
+        if (not all_lower) and der == 1:
+            return self._evaluate(x)
+
+        if der is None:
+            der = self.n
+
+        if all_lower and (x.size == 0 or self.r == 0):
+            cn = np.zeros((der, len(x), self.r), dtype=self.dtype)
+            return cn
+
+        if self._diff_cij is None:
+            # c[i,j] = xi[i] - xi[j]
+            c = self.xi[..., np.newaxis] - self.xi
+            # avoid division by 0 (diagonal entries are so far zero by construction)
+            np.fill_diagonal(c, 1)
+            # with np.errstate(divide='ignore', invalid='ignore'):
+            # c[i,j] = w[j] / (xi[i] - xi[j])
+            c = self.wi / c
+
+            # c[i,j] = (w[j] / w[i]) / (xi[i] - xi[j])
+            c = c / self.wi[..., np.newaxis]
+
+            # fill in correct diagonal entries: each column sums to 0
+            np.fill_diagonal(c, 0)
+            # c[j,j] = -sum_{i != j} c[i,j]
+            d = -c.sum(axis=1)
+            # c[i,j] = l_j(x_i)
+            np.fill_diagonal(c, d)
+
+            self._diff_cij = c
+
+        if self._diff_baryint is None:
+            # initialise and cache derivative interpolator and cijs;
+            # reuse weights (which depend only on interpolation points xi)
+            # to avoid unnecessary re-computation
+            self._diff_baryint = BarycentricInterpolator(xi=self.xi, yi=self._diff_cij @ self.yi, wi=self.wi)
+            self._diff_baryint._diff_cij = self._diff_cij
+
+        if all_lower:
+            # assemble matrix of derivatives from order 0 to order der-1,
+            # in the format required by _Interpolator1DWithDerivatives.
+            cn = np.zeros((der, len(x), self.r), dtype=self.dtype)
+            for d in range(der):
+                cn[d, :, :] = self._evaluate_derivatives(x, d+1, all_lower=False)
+            return cn
+
+        # recursively evaluate only the derivative requested
+        return self._diff_baryint._evaluate_derivatives(x, der-1, all_lower=False)
+
+
+def barycentric_interpolate(xi, yi, x, der=0, axis=0):
```

Comment:
Quick note about backward compatibility, we must observe the current order so add `der` after `axis`. Also we prefer to add keywords only arguments.
```suggestion
def barycentric_interpolate(xi, yi, x, axis=0, *, der=0):
```

Reply:
Unfortunately this brings the signature of `barycentric_interpolate` into conflict with `krogh_interpolate(xi, yi, x, der=0, axis=0)`, which I was hoping to avoid - is there a way to resolve this?

Reply:
So if I've understood correctly, the idea would be to:
- add a deprecation warning to `krogh_interpolate` if called with positional arguments for `der` or `axis`;
- add a deprecation warning to `barycentric_interpolate` if called with a positional argument for `axis`;
- allow both to take keyword arguments according to the new signature `*, der=0, axis=0)`.

If so that seems very sensible and I can add the deprecation warnings to the PR.

Reply:
Please don't do anything in this PR. This is something we would first need to discuss and plan for. Not that simple I am afraid and either way we would want to have dedicated PRs for that 😅

But yes that might look like similar to what you outlined.

Reply:
> Please don't do anything in this PR. This is something we would first need to discuss and plan for. Not that simple I am afraid and either way we would want to have dedicated PRs for that sweat_smile
> 
> But yes that might look like similar to what you outlined.

Agree let's stick with the suggested change for this PR otherwise it will hold up what is a nice enhancement. That process can be discussed/planned afterwards